File: packages/server/src/tcp_server.js
'use strict'

const randomstring = require('randomstring')
const net = require('net')
const uuid = require('uuid/v4')
const {
  arrayBufferWithMime,
} = require('arraybuffer-mime')
const { server } = require('./ws_server')
const { port, netPort } = require('./config')
const socks = require('./socks')
const { insertStreamLog } = require('./db')
const { isValidChannelName, normalizeChannel, isReservedKeyword } = require('./helpers')
const url = require('url')

process.setMaxListeners(0)

function genRandString() {
  return randomstring.generate({
    length: 6,
    charset: 'alphabetic',
    capitalization: 'lowercase',
    readable: true
  })
}

function channelTaken(s) {
  return !!socks[s]
}

function getRandomChannel() {
  let location = null

  // generate different path id if already being used
  do {
    const randString = genRandString()
    location = randString
  } while (socks[location] && !isValidChannelName(location))

  return location
}

function getShareUrl(address, port, channel) {
  address = address.split(':').splice(-1, 1)[0]

  if (!address || address === '1') {
    address = '127.0.0.1'
  }

  const hostUrl = process.env.HOST_URL || `http://${address}:${port}`
  const {protocol,host} = url.parse(hostUrl)
  let pathname = `s/${channel}`
  if (host === 'stream.ht') {
    pathname = channel
  }

  return `${protocol}//${host}/${pathname}`
}

const netConnections = {}

// netcat server
const netServer = net.createServer((socket) => {
  if (!socket.id) {
    socket.id = uuid()
  }

  let channel = null

  const info = socket.address()

  socket.pipe(socket)

  let line = 0
  let expired = false
  setTimeout(() => {
    expired = true

    if (!channel) {
      channel = getRandomChannel()
    }

    netConnections[channel] = socket
    const shareUrl = getShareUrl(info.address, port, channel)
    socket.write(`streamhut: streaming to ${shareUrl}\n\r`)
  }, 5)

  socket.on('data', async (buffer) => {
    if (line == 0 && !expired) {
      let data = buffer.toString()
      if (data[0] === '#') {
        const re = /#([a-zA-Z0-9]+)\n?\r?/
        const matches = data.match(re)
        if (matches.length > 1) {
          channel = normalizeChannel(matches[1])
          if (!isValidChannelName(channel)) {
            socket.write(`streamhut: channel name "${channel}" is not available`)
            socket.end()
            return
          }
          line++
          return
        }
      }
    }

    if (!channel) {
      return
    }

    const mime = 'shell'
    const abWithMime = arrayBufferWithMime(buffer.buffer, mime)
    insertStreamLog(channel, buffer)

    const clients = socks[channel]

    if (clients) {
      clients.forEach(client => {
        console.log(`streaming to ${client.id} ${channel}`)
        client.send(abWithMime)
      })
    }
  })
  .on('error', (error) => {
    console.error(error)
  })

})

function start() {
  netServer.listen({
    hostname: '0.0.0.0',
    port: netPort
  }, () => {
    console.log(`           TCP port: ${netPort}`)
  })
}

module.exports = { start }

File: pkg/tcpserver/tcpserver.go
package tcpserver

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"net"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/patrickmn/go-cache"
	gocache "github.com/patrickmn/go-cache"
	uuid "github.com/satori/go.uuid"
	"github.com/streamhut/streamhut/common/byteutil"
	"github.com/streamhut/streamhut/common/stringutil"
	common "github.com/streamhut/streamhut/common/util"
	"github.com/streamhut/streamhut/pkg/db"
	"github.com/streamhut/streamhut/pkg/util"
	"github.com/streamhut/streamhut/pkg/wsserver"
	"github.com/streamhut/streamhut/types"
)

// ErrInvalidQuotaLimit ...
var ErrInvalidQuotaLimit = fmt.Sprintf("Invalid quota size")

// ErrInvalidQuotaDuration ...
var ErrInvalidQuotaDuration = fmt.Sprintf("Invalid quota duration")

// DefaultBandwidthQuotaLimit ...
var DefaultBandwidthQuotaLimit = 1000 * 1000 * 10 // 10mb

// DefaultBandwidthQuotaDuration ...
var DefaultBandwidthQuotaDuration = 1 * time.Minute

// BandwidthQuotaLimit type
type BandwidthQuotaLimit uint64

// Server ...
type Server struct {
	host                   string
	listener               net.Listener
	port                   uint
	ws                     *wsserver.WS
	db                     db.DB
	shareBaseURL           string
	cache                  *gocache.Cache
	bandwidthQuotaLimit    BandwidthQuotaLimit
	bandwidthQuotaDuration time.Duration
}

// Config ...
type Config struct {
	Host                   string
	Port                   uint
	WS                     *wsserver.WS
	DB                     db.DB
	ShareBaseURL           string
	BandwidthQuotaLimit    uint64
	BandwidthQuotaDuration time.Duration
}

// NewServer ...
func NewServer(config *Config) *Server {
	shareBaseURL := config.ShareBaseURL
	if shareBaseURL != "" {
		if !strings.HasSuffix(shareBaseURL, "/") {
			shareBaseURL += "/"
		}
	}

	if config.BandwidthQuotaLimit > 0 && config.BandwidthQuotaDuration == 0 {
		log.Fatal(ErrInvalidQuotaDuration)
	}

	if config.BandwidthQuotaDuration > 0 && config.BandwidthQuotaLimit == 0 {
		log.Fatal(ErrInvalidQuotaLimit)
	}

	return &Server{
		host:                   config.Host,
		port:                   config.Port,
		ws:                     config.WS,
		db:                     config.DB,
		shareBaseURL:           shareBaseURL,
		cache:                  cache.New(gocache.DefaultExpiration, gocache.DefaultExpiration),
		bandwidthQuotaLimit:    BandwidthQuotaLimit(config.BandwidthQuotaLimit),
		bandwidthQuotaDuration: config.BandwidthQuotaDuration,
	}
}

// Start ...
func (s *Server) Start() error {
	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", s.host, s.port))
	if err != nil {
		return err
	}

	defer listener.Close()
	s.listener = listener

	for {
		conn, err := listener.Accept()
		if err != nil {
			return err
		}

		id := uuid.Must(uuid.NewV4(), nil)
		client := &wsserver.Conn{
			Channel: "",
			Netconn: conn,
			ID:      id.String(),
		}

		go s.handleRequest(client)
	}
}

func (s *Server) randChannel() string {
	for {
		channel := stringutil.RandStringRunes(6)
		_, ok := s.ws.Socks[channel]
		if !ok && common.ValidChannelName(channel) {
			return channel
		}
	}
}

func (s *Server) channelTaken(channel string) bool {
	_, ok := s.ws.Socks[channel]
	return ok
}

func (s *Server) handleRequest(client *wsserver.Conn) {
	reader := bufio.NewReader(client.Netconn)
	index := 0
	channelReadExpired := false

	var ip string
	if addr, ok := client.Netconn.RemoteAddr().(*net.TCPAddr); ok {
		ip = addr.IP.String()
	}

	// NOTE: a timeout to allow reading of channel first
	time.AfterFunc(5*time.Millisecond, func() {
		channelReadExpired = true
		if client.Channel == "" {
			client.Channel = s.randChannel()
		}
		//s.ws.netConnections[client.channel] = append(s.ws.Socks[client.channel], client)
		str := fmt.Sprintf("streamhut: streaming to %s\n", s.shareURL(client))
		client.Netconn.Write([]byte(str))
	})

	for {
		line := make([]byte, 1024)
		n, err := reader.Read(line)
		switch err {
		case nil:
			if s.bandwidthQuotaLimit > 0 {
				_, exp, found := s.cache.GetWithExpiration(ip)
				if found {
					expiresIn := time.Duration(exp.Unix()-time.Now().Unix()) * time.Second

					if expiresIn.Seconds() == 0 {
						client.ResetBandwidthQuota()
						s.cache.Delete(ip)
					} else {
						msg := fmt.Sprintf("streamhut: bandwidth quota reached. Try again in %vs", expiresIn.Seconds())
						log.Printf("quota reached for ip %v; can retry in %vs\n", ip, expiresIn.Seconds())
						client.Netconn.Write([]byte(msg))

						// NOTE: timeout must be less than channelReadExpired
						time.Sleep(4 * time.Millisecond)
						client.Netconn.Close()
						return
					}
				}

				client.TollBandwidth(uint64(n))
				if client.BandwidthQuotaUsed() > s.bandwidthQuotaLimit.Uint64() {
					expiresIn := time.Duration(int(s.bandwidthQuotaDuration.Seconds())-time.Now().Second()) * time.Second
					s.cache.Set(ip, time.Now(), expiresIn)
					msg := fmt.Sprintf("\nstreamhut: bandwidth quota reached. Try again in %vs", expiresIn.Seconds())
					log.Printf("quota reached for ip %v; can retry in %vs\n", ip, expiresIn.Seconds())
					client.Netconn.Write([]byte(msg))
					time.Sleep(1 * time.Second)
					client.Netconn.Close()
					return
				}
			}

			// echo back to client
			client.Netconn.Write(line)
		case io.EOF:
			return
		default:
			os.Exit(0)
		}

		if index == 0 && !channelReadExpired {
			if len(line) > 0 && line[0] == '#' {
				re := regexp.MustCompile(`#([a-zA-Z0-9]+)\n?\r?`)
				matches := re.FindAllStringSubmatch(string(line), -1)
				if len(matches) > 0 && len(matches[0]) > 1 {
					client.Channel = common.NormalizeChannelName(matches[0][1])
					if !common.ValidChannelName(client.Channel) {
						msg := fmt.Sprintf("streamhut: channel name %q is not available", client.Channel)
						client.Netconn.Write([]byte(msg))
						if err := client.Netconn.Close(); err != nil {
							log.Fatal(err)
						}

						continue
					}

					index++
					continue
				}
			}
		}

		if client.Channel == "" {
			continue
		}

		mime := "shell"
		bufferWithMime := byteutil.BufferWithMime(line, mime)

		go s.db.InsertStreamLog(&types.StreamLog{
			Channel: client.Channel,
			Data:    line,
		})

		clients, ok := s.ws.Socks[client.Channel]
		if ok {
			for _, cl := range clients {
				fmt.Sprintf("streaming to %s on channel %q", cl.ID, cl.Channel)
				if cl.Netconn != nil {
					if _, err := cl.Netconn.Write(bufferWithMime); err != nil {
						log.Fatal(err)
					}
				}

				if cl.Wsconn != nil {
					if err = cl.Write(bufferWithMime); err != nil {
						log.Fatal(err)
					}
				}
			}
		}
	}
}

func (s *Server) shareURL(client *wsserver.Conn) string {
	if s.shareBaseURL != "" {
		return fmt.Sprintf("%s%s", s.shareBaseURL, client.Channel)
	}

	host := s.host
	if host == "" {
		host = "127.0.0.1"
	}

	hostURL := fmt.Sprintf("http://%s:%d", host, s.port)
	u, err := url.Parse(hostURL)
	if err != nil {
		log.Fatal(err)
	}

	host = u.Host
	protocol := u.Scheme
	pathname := fmt.Sprintf("s/%s", client.Channel)
	if !strings.HasSuffix(u.Path, "/") {
		pathname = "/" + pathname
	}

	pathname = u.Path + pathname
	return fmt.Sprintf("%s//%s%s", protocol, host, pathname)
}

// Port ...
func (s *Server) Port() uint {
	return s.port
}

// BandwidthQuotaLimit ...
func (s *Server) BandwidthQuotaLimit() BandwidthQuotaLimit {
	return s.bandwidthQuotaLimit
}

// BandwidthQuotaDuration ...
func (s *Server) BandwidthQuotaDuration() time.Duration {
	return s.bandwidthQuotaDuration
}

// BandwidthQuotaEnabled ...
func (s *Server) BandwidthQuotaEnabled() bool {
	return s.bandwidthQuotaLimit > 0
}

// Uint64 ...
func (b BandwidthQuotaLimit) Uint64() uint64 {
	return uint64(b)
}

// String ...
func (b BandwidthQuotaLimit) String() string {
	return util.Uint64ToStorageSizeString(uint64(b))
}
